src/core/control/zoom/ZoomControl.cpp

```diff
@@ -23,7 +23,7 @@ auto onScrolledwindowMainScrollEvent(GtkWidget* widget, GdkEventScroll* event, Z
     guint state = event->state & gtk_accelerator_get_default_mod_mask();
 
     // do not handle e.g. ALT + Scroll (e.g. Compiz use this shortcut for setting transparency...)
-    if (state != 0 && (state & ~(GDK_CONTROL_MASK | GDK_SHIFT_MASK))) {
+    if (as_signed(state) & ~(GDK_CONTROL_MASK | GDK_SHIFT_MASK)) {
```

Comment:
Actually the state and  `GDK_CONTROL_MASK` and `GDK_SHIFT_MASK` have matching types. The thing is that the bitwise complement operator first makes an integer promotion of `GDK_CONTROL_MASK | GDK_SHIFT_MASK` and thus turns it into a signed integer. So your suggestion doesn't work.

Reply:
Ok, I got this slightly wrong. Nevertheless, if those constants were `uint16_t`, then bitwise complement would turn them signed. This is kind of surprising, isn't it?

```c++
#include <iostream>

int main()
{
    uint16_t a = 1;
    std::cout << ~a << std::endl;
    if (~a < 0) {
        std::cout << "See, bitwise complement turned it to a signed integer";
    }

    return 0;
}
```

Reply:
> The problem here is that a static_cast (on either side) could modify the bits in theory. I'm guessing there are no infrastructures on which this actually does not work, but I still think this is UB.

A static cast between 2 integral types of the same size does not alter the bit representation (since c++20's 2's compliment enforcement), but of course it would be more descriptive, to `bit_cast` them. So using my as_unsigned or as_signed is always safe (Unless the value represented is not in the value range of both types (until c++17)).

>  if those constants were uint16_t, then bitwise complement would turn them signed. This is kind of surprising, isn't it?

The problem here is that CPP/C unfortunately does integer promotion https://en.cppreference.com/w/c/language/conversion#Integer_promotions. I think this is because most machines define machine code for 32-bit integer operations, and the minimum register size in an ALU is then 32 bit. Therefore, any type which has sizeof(T) < sizeof(int) is promoted when you copy it into a register.

The cleanest way would be, to define custom operators (^~&|) for such enums or wrap them into a helper class, which is something like BitField<MyEnum> and define those operators for them.

To get the possibly promoted type for an operation is:

`decltype( x op y )` like `decltype( x << y )` 

For enums in general, we might want to use something like that: 
```
EnumV operator^(EnumV a, EnumV b) {
    using UnsignedT = std::make_unsigned_t<std::std::underlying_type_t<EnumV>>; // To strip of bits after promotion
    auto value = as_unsigned(as_unsigned(std::to_underlying(a)) ^ as_unsigned(std::to_underlying(a)));  // ensure type is unsigned even after promotion to signed value.
    return static_cast<EnumV>(static_cast<UnsignedT>(value)); // Strip of bits in case of promotion and return Enum type.
}
```